src/Database/Relations/BelongsOrMorphTo.php

```diff
@@ -0,0 +1,87 @@
+<?php namespace October\Rain\Database\Relations;
+
+trait BelongsOrMorphTo
+{
+    /**
+     * Associate the model instance to the given parent.
+     *
+     * @param  \Illuminate\Database\Eloquent\Model  $model
+     * @return \Illuminate\Database\Eloquent\Model
+     */
+    public function associate($model)
+    {
+        /**
+         * @event model.relation.beforeAssociate
+         * Called before associating a relation to the model (only for BelongsTo/MorphTo relations)
+         *
+         * Example usage:
+         *
+         *     $model->bindEvent('model.relation.beforeAssociate', function (string $relationName, \October\Rain\Database\Model $relatedModel) use (\October\Rain\Database\Model $model) {
+         *         if ($relationName === 'dummyRelation') {
+         *             throw new \Exception("Invalid relation!");
+         *         }
+         *     });
+         *
+         */
+        $this->parent->fireEvent('model.relation.beforeAssociate', [$this->relationName, $this->related]);
+
+        $result = parent::associate($model);
+
+        /**
+         * @event model.relation.afterAssociate
+         * Called after associating a relation to the model (only for BelongsTo/MorphTo relations)
+         *
+         * Example usage:
+         *
+         *     $model->bindEvent('model.relation.afterAssociate', function (string $relationName, \October\Rain\Database\Model $relatedModel) use (\October\Rain\Database\Model $model) {
+         *         $relatedClass = get_class($relatedModel);
+         *         traceLog("Relation {$relationName} was associated to model {$relatedClass}.");
+         *     });
+         *
+         */
+        $this->parent->fireEvent('model.relation.afterAssociate', [$this->relationName, $this->related]);
```

Comment:
All of the after events should include the result, you might have to have it as the third argument so that you can support the existing relation event.

Reply:
@LukeTowers In this case, the parent returns $this->child which is the same model on which the event is fired, so I don't think this is very helpful, is it?

Reply:
I suppose not, $this->child is equal to $this->related?

Reply:
@LukeTowers No, it's equal to `$this->parent` (for BelongsTo and MorphTo relations)

Reply:
Ah, so the return value of this method is the parent model that the event is bound to in the first place?

Reply:
Yop.